packages/cdktf-cli/lib/project-execution.ts

```diff
@@ -0,0 +1,448 @@
+  planNeedsNoApply: (
+    _context: ProjectContext,
+    event: { data: TerraformPlan }
+  ) => event.data.needsApply === false,
+};
+
+const projectExecutionMachine = createMachine<ProjectContext, ProjectEvent>(
+  {
+    id: "project",
+    initial: "idle",
+    context: {
+      onProgress: () => {}, // eslint-disable-line @typescript-eslint/no-empty-function
+      // Mandantorily set by withContext
+      targetAction: "synth",
+      synthCommand: "",
+      workingDirectory: "",
+      outDir: "",
+    },
+    states: {
+      idle: {
+        on: {
+          START: {
+            target: "synth",
+            actions: assign({
+              targetAction: (_context, event) => event.targetAction,
+              targetStack: (_context, event) => event.targetStack,
+            }),
+          },
+        },
+      },
+      synth: {
+        invoke: {
+          id: "synth",
+          src: "synthProject",
+          onError: {
+            target: "error",
+            actions: assign({
+              message: (_context, event) => event.data,
+            }),
+          },
+          onDone: [
+            {
+              target: "done",
+              actions: assign({
+                synthesizedStacks: (_context, event) => event.data,
+              }),
+              cond: {
+                type: "onTargetAction",
+                name: "actionIsSynth",
+                value: "synth",
+              },
+            },
+            {
+              target: "gatherOutput",
+              actions: assign({
+                synthesizedStacks: (_context, event) => event.data,
+              }),
+              cond: {
+                type: "onTargetAction",
+                name: "actionIsOutput",
+                value: "output",
+              },
+            },
+            {
+              target: "diff",
+              actions: assign({
+                synthesizedStacks: (_context, event) => event.data,
+              }),
+            },
+          ],
+        },
+      },
+      diff: {
+        invoke: {
+          id: "diff",
+          src: "diff",
+          onError: {
+            target: "error",
+            actions: assign({
+              message: (_context, event) => {
+                const errorMessage = event.data.stderr || event.data;
+                return `terraform plan errored with: \n${errorMessage}`;
+              },
+            }),
+          },
+          onDone: [
+            {
+              target: "done",
+              actions: assign({
+                targetStackPlan: (_context, event) => event.data,
+              }),
+              cond: {
+                type: "onTargetAction",
+                name: "actionIsDiff",
+                value: "diff",
+              },
+            },
+
+            {
+              actions: assign({
+                targetStackPlan: (_context, event) => event.data,
+              }),
+              target: "gatherOutput",
+              cond: "planNeedsNoApply",
```

Comment:
Why is a diff printing outputs in this case but not in the case above? Is there any magic to `planNeedsNoApply` (such as only being present if a deploy was the goal and absent in case of a plan?)

Reply:
Ah, wait. As the above case catches the diff target, it will end there (without printing outputs).
This will hence only print outputs if the target was not diff.
Can we make this more clear?

Does it make sense to run `gatherOutput` if the target is `destroy`? (It would run that, wouldn't it?)

Reply:
Ah, btw: You could also make deploy ignore a plan that needs no deploy (no-op) and then remove this check. I don't know which strategy is better, however. No idea at all, might as well stay as is. State Machine would be less complicated but also less explicit 🤷🤷🤷🤷

Reply:
From my point of view all "workflow" related logic should live in the state machine. I'd say the functions we call have too much logic right now, they should be as dumb as possible. But I didn't want to overstretch it, I rather look at it when we work on the upcoming features